src/analyses/basePriv.ml

```diff
@@ -0,0 +1,1664 @@
+open Prelude.Ana
+open Analyses
+open GobConfig
+open BaseUtil
+module Q = Queries
+
+module IdxDom = ValueDomain.IndexDomain
+
+module VD     = BaseDomain.VD
+module CPA    = BaseDomain.CPA
+module BaseComponents = BaseDomain.BaseComponents
+
+
+module type S =
+sig
+  module D: Lattice.S
+  module G: Lattice.S
+
+  val startstate: unit -> D.t
+
+  val read_global: Q.ask -> (varinfo -> G.t) -> BaseComponents (D).t -> varinfo -> VD.t
+  val write_global: Q.ask -> (varinfo -> G.t) -> (varinfo -> G.t -> unit) -> BaseComponents (D).t -> varinfo -> VD.t -> BaseComponents (D).t
+
+  val lock: Q.ask -> (varinfo -> G.t) -> BaseComponents (D).t -> LockDomain.Addr.t -> BaseComponents (D).t
+  val unlock: Q.ask -> (varinfo -> G.t) -> (varinfo -> G.t -> unit) -> BaseComponents (D).t -> LockDomain.Addr.t -> BaseComponents (D).t
+
+  val sync: Q.ask -> (varinfo -> G.t) -> BaseComponents (D).t -> [`Normal | `Join | `Return | `Init | `Thread] -> BaseComponents (D).t * (varinfo * G.t) list
+
+  val escape: Q.ask -> (varinfo -> G.t) -> (varinfo -> G.t -> unit) -> BaseComponents (D).t -> EscapeDomain.EscapedVars.t -> BaseComponents (D).t
+  val enter_multithreaded: Q.ask -> (varinfo -> G.t) -> (varinfo -> G.t -> unit) -> BaseComponents (D).t -> BaseComponents (D).t
+  val threadenter: Q.ask -> BaseComponents (D).t -> BaseComponents (D).t
+
+  (* TODO: better name *)
+  val is_private: Q.ask -> varinfo -> bool
```

Comment:
What is the purpose of this now?

Reply:
I keep asking myself this question each time I see this function. Especially since it makes no sense for most implementations. But I'm also not sure about the implications of removing it since there is one use in `base.ml`:
https://github.com/goblint/analyzer/blob/4b0fa27033db36447f04b5b016c61ce8b26cc88e/src/analyses/base.ml#L1042-L1048

I think it somehow matters for protection-based privatizations (or maybe just the `sync`-based ones like `old` and `global-vesal`?) because it wants to avoid spurious side-effects when setting in `invariant`. For others we've already established that `invariant` would have to cause the thread to write and this isn't used I guess.

Maybe we should just make `effect` or something similar to be the argument of `Priv.write_global` so it would be internal to a `Priv` what to do about it.

Reply:
> Maybe we should just make effect or something similar to be the argument of Priv.write_global so it would be internal to a Priv what to do about it.

Yes, I think that sounds reasonable!

Reply:
I remembered another worry I had with `is_private`. It used to be used for `make_entry`:
https://github.com/goblint/analyzer/blob/85536ab5c9cc0ffe5088af1c01eef34be27fa00e/src/analyses/base.ml#L1834

At some point I just removed that extra filtering on `traces` to get something else working and probably didn't see a change on the regression tests, but there might've been a reason for that use of `is_private` as well.

Reply:
Anyway, the `make_entry` use case wouldn't be enough of a reason to keep `is_private` around anyway, because the later change of making `make_entry` behave differently for `threadenter` via `Priv` should allow handling whatever the intended use case was internally in the `Priv`, just like the extra argument of `write_global` does, so I'm just going to make this change.

The case for the old privatization and `invariant` was even more subtle: `write_global` there doesn't `sideg`, but just adds to `CPA` and only the following `sync` does the side effecting. So `invariant` prevented the `CPA` change under some conflicting conditions where we just traced?